### ISSUE TITLE
Read only fields

### ISSUE THREAD

**Issue by andgith** [laravel/nova-issues#19]
Are computed fields able to be displayed in the detail panel? Or is there a way to add a read only field to the resource detail?

**Comment by cotton9** [comment]
Computed fields are already visible within the resource "Detail" view.  Perhaps you meant "Update" views?  I too am trying to figure out how to display fields as readonly on update forms.

**Comment by andgith** [comment]
Ah yes, my mistake. Read only fieldsin the update/edit view would help give context in some situations.

**Comment by WaveHack** [comment]
I'd also like to see readonly fields, to be used for example with auto-generated, immutable and saved fields (e.g. URL slugs).

Or to read only for example a JSON settings field tied to a User, where I don't want to edit JSON by hand by accident.

**Comment by gms8994** [comment]
@davidhemphill is this moving to a different type of issue?

**Comment by davidhemphill** [comment]
Nah, I closed a couple to get them out of the main view since they'll be active so long. But I'm just gonna leave them open.

**Comment by hariesnurikhwan** [comment]
Is there any workaround/hack for this?

**Comment by roshangautam** [comment]
If you want to exclude a field from `PUT` request on update, just add an empty fillUsing callback on that field as follows:

        Text::make("IgnoreMePlease")
                ->fillUsing(function() {
                    return;
                })

Hope this helps.

**Comment by ashbakernz** [comment]
Yep I agree this is a must have feature 👌

**Comment by vtamm** [comment]
+1

Would very much appreciate the ability to disable fields!

**Comment by RicLeP** [comment]
This is essential. Some fields shouldn’t be edited here as are dynamically created elsewhere in the app. Same does for read only resources #426

**Comment by sonalmahajan01** [comment]
+1

**Comment by sergionader** [comment]
Hey,

Having the ability to show a field disable during editing is something that would help a lot. 

Let's say you have a status field that changes by any background process and that is not intended to be changed by the user.

While editing a record, the user may want to know its status, though she can't change it. It sounds as a pretty standard need. 

Some as simple as:
`text::make('status')->disabled()` or any of the other variations shown on this thread. 

Thank you!

**Comment by dillingham** [comment]
@davidhemphill maybe something like..

```php
->withMeta('inputAttr', ['maxlength' => 5])
->withMeta('inputAttr', ['disabled' => true])
->withMeta('inputAttr', ['readonly' => true])
->withMeta('inputAttr', ['class' => 'border-b bg-50'])
```

And `v-bind="inputAttr"`

**Comment by davidhemphill** [comment]
The ability to set any attribute on a text or textarea field is coming in the next release. You'll be able to specify something like this:

```php
Text::make('Name')
    ->withMeta(['extraAttributes' => [
        // 'placeholder' => '0',
        // 'readonly' => true,
        // 'disabled' => true,
        // 'type' => 'number',
        // 'step' => 4,
        // 'required' => true,
    ]]),
```

**Comment by chrisbjr** [comment]
> The ability to set any attribute on a text or textarea field is coming in the next release. You'll be able to specify something like this:
> 
> ```
> Text::make('Name')
>     ->withMeta(['extraAttributes' => [
>         // 'placeholder' => '0',
>         // 'readonly' => true,
>         // 'disabled' => true,
>         // 'type' => 'number',
>         // 'step' => 4,
>         // 'required' => true,
>     ]]),
> ```

Thanks! When is the next release coming?

**Comment by noahlocke** [comment]
@davidhemphill running 1.1.4, but not seeing these changes taking affect:

```php
            Text::make('Provider')->withMeta([
                'extraAttributes' => [
                    'placeholder' => 'something'
                ]
            ]),
```

**Comment by davidhemphill** [comment]
Make sure you republished your assets with `artisan nova:publish --force` to be sure.

**Comment by approached** [comment]
Same problem. It's not working. :(

**Comment by approached** [comment]
Same code...
```php
Text::make('Provider')->withMeta([
                'extraAttributes' => [
                    'placeholder' => 'something'
                ]
            ]),
```

**Comment by approached** [comment]
Json Response looks good:
```
{
component: "text-field",
prefixComponent: true,
indexName: "Provider",
name: "Provider",
attribute: "provider",
value: null,
panel: null,
sortable: false,
textAlign: "left",
extraAttributes: {
  placeholder: "something"
}
```

**Comment by cjf** [comment]
I was trying this out to set the "disabled" attribute, but it seems to not work on Select fields.  For example, this works...

`Text::make('Type', 'type')
            ->withMeta([
                'extraAttributes' => [
                    'disabled' => true,
                ]
            ]),`

But this does not...

`Select::make('Type', 'type')->options([
                '0' => 'A',
                '1' => 'B',
                '2' => 'C',
            ])->withMeta([
                'extraAttributes' => [
                    'disabled' => true,
                ]
            ])`

The latter produces this code...

`<select id="type" class="w-full form-control form-select">
<option value="" selected="selected" disabled="disabled">
                Choose an option
            </option> <option value="0">
                A
            </option><option value="1">
                B
            </option><option value="2">
                C
            </option></select>`

Basically, it puts the disabled attribute in the wrong place.

**Comment by dillingham** [comment]
@cjf extraAttributes was only added to text and textarea

**Comment by cjf** [comment]
@dillingham Is it going to be implemented on other field types?

**Comment by noahlocke** [comment]
@davidhemphill I did run that command, `artisan nova:publish --force` (though I hope I shouldn't have to nuke all of my views?) Still nothing. I can see  `extraAttributes` on the Vue component:

<img width="290" alt="screen shot 2018-10-02 at 10 55 46 pm" src="https://user-images.githubusercontent.com/6596677/46389245-78dddc80-c696-11e8-8864-e10f8e9899a4.png">

**Comment by davidhemphill** [comment]
Honestly seems like your javascript is still cached. I don't know how to diagnose this any further; we haven't seen any other reports of it not working.

**Comment by jasonlav** [comment]
Disabled works on text strings, but not select menus (and apparently some other fields). This is absolutely a bug; this is not a request nor intentional behavior.

**Comment by fbingha** [comment]
Agree with @metrixinfo 

I have BelongsTo::make('User') and I don't want a select field with thousdands of users to appear on the edit form. Just showing the current user, unable to change is what I want.

**Comment by ashbakernz** [comment]
> It does not seem to work with a BelongsTo relationship.
> 
> BelongsTo::make('User'),
> 
> It allows the user to be changed when editing the field. We have close to a million users and this is for a 1:1 relationship table that that must never be changed.
> 
> However I like the idea of showing the user the entry belongs to when editing it.

I've come across the same problem. I personally like the syntax ```BelongsTo::make('User')->isReadOnly()```

**Comment by Milkhan** [comment]
Can we give disabled attribute to an option of a select element ? Or seperator maybe?
 'MB' => 'Manitoba',
  '--' => '-------------', disabled ,

**Comment by jfelixoptimiumnc** [comment]
Hey, just found that work, kinda weird hu:          

   ((Auth::user()->profil == "admin")) ?
                Currency::make('Amount')
                    ->hideFromIndex()
            :
                Currency::make('Amount')
                    ->exceptOnForms()
                    ->hideFromIndex()
            ,

**Comment by notflip** [comment]
> > It does not seem to work with a BelongsTo relationship.
> > BelongsTo::make('User'),
> > It allows the user to be changed when editing the field. We have close to a million users and this is for a 1:1 relationship table that that must never be changed.
> > However I like the idea of showing the user the entry belongs to when editing it.
> 
> I've come across the same problem. I personally like the syntax `BelongsTo::make('User')->isReadOnly()`

Is this something that is already implemented at this time? Or how are you solving this now?

**Comment by yuqinghong** [comment]
> Agree with @metrixinfo
> 
> I have BelongsTo::make('User') and I don't want a select field with thousands of users to appear on the edit form. Just showing the current user, unable to change is what I want.

if you don't want to edit it , you can let not show on form, so you can try `BelongsTo::make('user')->exceptOnForms()` 
如果想使该字段为只读， 那不要在表单里显示该字段好啦！可以试试 `BelongsTo::make('user')->exceptOnForms()`

**Comment by christianmarth** [comment]
This still doesn't work on select fields so needs to be addressed

**Comment by subangkit** [comment]
@davidhemphill , I think this is should add class "disabled"

**Comment by gustav1105** [comment]
Read only does not work on belongsTo , not using plugin
